proposals/format.md

```diff
@@ -0,0 +1,320 @@
+as it becomes an implementation detail of other parts of the proposal. 
+
+### Efficient interpolated strings
+Interpolated strings are a popular yet innefecient feature in C#. The most common syntax, using an interpolated `string`
+as a `string`, translates into a `string.Format(string, params object[])` call. That will inccur boxing allocations for 
+all value types, intermediate `string` allocations as the implementation largely uses `object.ToString` for formatting
+as well as array allocations once the number of arguments exceeds the amount of parameters on the "fast" overloads of 
+`string.Format`. 
+
+The language will change it's interpolation lowering to consider alternate overloads of `string.Format`. It will
+consider all forms of `string.Format(string, params)` and pick the "best" overload which satisfies the argument types.
+The "best" `params` overload will be determined by the rules discussed above. This means interpolated `string` can now
+bind to very efficient overloads like `string.Format(string format, params ReadOnlySpan<Variant> args)`. In many cases
+this will remove all intermediate allocations.
+
+### Customizable interpolated strings
+Developers are able to customize the behavior of interpolated strings with `FormattableString`. This contains the data
+which goes into an interpolated string: the format `string` and the arguments as an array. This though still has the 
+boxing and argument array allocation as well as the allocation for `FormattableString` (it's an `abstract class`). Hence
+it's of little use to applications which are allocation heavy in `string` formatting.
+
+To make interopolated string formatting efficient the language will recognize a new type: 
+`System.ValueFormattableString`. All interpolated strings will have a target type conversion to this type. This will 
+be implemented by translating the interpolated string into the call `ValueFormattableString.Create` exactly as is done
+for `FormattableString.Create` today. The language will support all `params` options described in this document when
+looking for the most suitable `ValueFormattableString.Create` method. 
+
+``` csharp
+readonly struct ValueFormattableString {
+    public static ValueFormattableString Create(Variant v) { ... } 
+    public static ValueFormattableString Create(string s) { ... } 
+    public static ValueFormattableString Create(string s, params ReadOnlySpan<Variant> collection) { ... } 
+}
+
+class ConsoleEx { 
+    static void Write(ValueFormattableString f) { ... }
+}
+
+class Program { 
+    static void Main() { 
+        ConsoleEx.Write(42);
+        ConsoleEx.Write($"hello {DateTime.UtcNow}");
+
+        // Translates into 
+        ConsoleEx.Write(ValueFormattableString.Create((Variant)42));
+        ConsoleEx.Write(ValueFormattableString.Create(
+            "hello {0}", 
+            VariantCollection.Create((Variant)DateTime.UtcNow));
```

Comment:
this approach seems unfortunate... because won't i now have to figure out how to break `"hello {0}"` up, even thouh the language already did that for me?  I'm coming from this from a JS/TS perspective where this is the natural way this data is packaged up and passed along to template processors.

Reply:
This is the approach that interpolated strings use today. The goal of this proposal is to make the existing process more efficient vs. opening up the door for text templating. That was brought up though in LDM and we're going to look into it a bit though. If you have some pointers to the JS / TS similar features I'd like to read up on them.

Reply:
> This is the approach that interpolated strings use today

That's fair.  Though it would be nice if both forms could be supported.  Having to have clients manually parse strings is a PITA.   JS/TS info on template strings is here:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Importantly: 

```
function template(strings, ...keys) {
  return (function(...values) {
    var dict = values[values.length - 1] || {};
    var result = [strings[0]];
    keys.forEach(function(key, i) {
      var value = Number.isInteger(key) ? values[key] : dict[key];
      result.push(value, strings[i + 1]);
    });
    return result.join('');
  });
}
```

This is how one could write a function that accepts an interpolated string.  From TypeScript's perspective, the signature is:

```ts
function template(literals: TemplateStringsArray, ...placeholders: string[]) {
```

TmplateStringsArray is also super interesting.  It is typed as:


```ts
interface TemplateStringsArray extends ReadonlyArray<string> {
    readonly raw: ReadonlyArray<string>;
}
```

Meaning, it actually represents *two* arrays of strings.  For example if i had:

```whatever `this is a \t { template } \t string literal`; ```

Then the first arg passed to 'whatever' would look like this:

```js
{
     [ "this is a \t ", " \t string literal"]
     raw: [ "this is a \\t ", " \\t string literal" ]
}
```

So JS gives both the original *raw* strings and the interpreted language strings.  I'm not asking for that myself, but at least having the separate strings makes it so much easier to do things.